compiler/rustc_resolve/src/late.rs

```diff
@@ -1539,8 +1539,18 @@ impl<'a, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
                         .unwrap_or_else(|| self.fresh_binding(ident, pat.id, pat_src, bindings));
                     self.r.record_partial_res(pat.id, PartialRes::new(res));
                 }
-                PatKind::TupleStruct(ref path, ..) => {
-                    self.smart_resolve_path(pat.id, None, path, PathSource::TupleStruct(pat.span));
+                PatKind::TupleStruct(ref path, ref sub_patterns) => {
+                    let pattern_spans: Vec<Span> =
+                        sub_patterns.iter().map(|p| p.span.clone()).collect();
+                    self.smart_resolve_path(
+                        pat.id,
+                        None,
+                        path,
+                        PathSource::TupleStruct(
+                            pat.span,
+                            self.r.arenas.alloc_pattern_spans(&pattern_spans),
```

Comment:
```suggestion
                            self.r.arenas.alloc_pattern_spans(sub_patterns.iter().map(|p| p.span)),
```
You can feed an iterator directly to `alloc_pattern_spans`.